exchange/exchange.go

```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
 			}
 			targData.setTargeting(auc, bidRequest.App != nil, bidCategory)
 		}
+
+		if requestExt.Prebid.SupportDeals {
+			dealErrs := applyDealSupport(bidRequest, auc)
+			if len(dealErrs) > 0 {
+				errs = append(errs, dealErrs...)
+			}
+		}
 	}
 
 	// Build the response
 	return e.buildBidResponse(ctx, liveAdapters, adapterBids, bidRequest, resolvedRequest, adapterExtra, auc, errs)
 }
 
+type DealTierInfo struct {
+	Prefix      string `json:"prefix"`
+	MinDealTier int    `json:"minDealTier"`
+}
+
+type DealTier struct {
+	Info *DealTierInfo `json:"dealTier,omitempty"`
+}
+
+type BidderDealTier struct {
+	DealInfo map[string]*DealTier
+}
+
+// Updates targeting keys with deal prefixes if minimum deal tier exceeded
+func applyDealSupport(bidRequest *openrtb.BidRequest, auc *auction) []error {
+	errs := []error{}
+	impDealMap := getDealTiers(bidRequest)
+
+	for impID, topBidsPerImp := range auc.winningBidsByBidder {
+		impDeal := impDealMap[impID].DealInfo
+		for bidderName, topBidPerBidder := range topBidsPerImp {
+			bidderString := bidderName.String()
+
+			if topBidPerBidder.dealPriority > 0 {
+				if validateDealTier(impDeal[bidderString]) {
+					updateCatDur(topBidPerBidder, impDeal[bidderString].Info)
+				} else {
+					errs = append(errs, fmt.Errorf("dealTier configuration invalid for bidder '%s', imp ID '%s'", bidderString, impID))
+				}
+			}
+		}
+	}
+
+	return errs
+}
+
+// Creates map of impression to bidder deal tier configuration
+func getDealTiers(bidRequest *openrtb.BidRequest) map[string]*BidderDealTier {
+	impDealMap := make(map[string]*BidderDealTier)
+
+	for _, imp := range bidRequest.Imp {
+		var bidderDealTier BidderDealTier
+		err := json.Unmarshal(imp.Ext, &bidderDealTier.DealInfo)
+		if err != nil {
+			continue
+		}
+
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
+	if impDeal.Info != nil {
+		// Remove whitespace from prefix before checking if it can be used
+		impDeal.Info.Prefix = regexp.MustCompile(`\s+`).ReplaceAllString(impDeal.Info.Prefix, "")
+		return len(impDeal.Info.Prefix) > 0 && impDeal.Info.MinDealTier > 0
+	}
+	return false
+}
+
+func updateCatDur(bid *pbsOrtbBid, dealTierInfo *DealTierInfo) {
```

Comment:
This is likely industry knowledge I'm lacking, but what is a 'CatDur'?

Reply:
'CatDur' is a shortening of the key value `hb_pb_cat_dur` but I don't think it is a standard shortening. I will update this function to be `updateHbPbCatDur` so that it is more specific.

Reply:
Thank you. That makes sense.